### PR TITLE
add extension "underspeech"

### PR DESCRIPTION
**Pull request by PuzzlingGGG**
underspeech basically adds like undertale like speaking

### CODE REVIEW

**Comment by PuzzlingGGG** [comment]
image
this is what it looks like

**Comment by lego7set** [comment]
nice extension! also i think you should use Scratch.(vm.)renderer.canvas for getting the canvas

**Comment by lego7set** [comment]
consider adding a catch branch to your showDialogueFromJSON func

File: static/extensions/PuzzlingGGG/underspeech.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
+class Underspeech {
+  constructor() {
+    this.dialogueContainer = this.createDialogueContainer();
+    const canvas = document.getElementById('sulfurous') || document.querySelector('canvas');
```
Comment (SharkPool-SP) [inline static/extensions/PuzzlingGGG/underspeech.js:4]: use `Scratch.vm.renderer.canvas` to get the canvas

File: static/extensions/PuzzlingGGG/underspeech.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
+class Underspeech {
```
Comment (SharkPool-SP) [inline static/extensions/PuzzlingGGG/underspeech.js:1]: wrap the extension in `(function (Scratch) {/*code here*/})(Scratch);`

File: static/extensions/PuzzlingGGG/underspeech.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                            defaultValue: 'normal'
+                        },
+                    }
+                },
+                {
+                    opcode: 'hideDialogue',
+                    blockType: Scratch.BlockType.COMMAND,
+                    text: 'hide dialogue'
+                },
+                '---','---','---',
```
Comment (SharkPool-SP) [inline static/extensions/PuzzlingGGG/underspeech.js:42]: you dont need 3 '---'

It does/adds nothing

File: static/extensions/PuzzlingGGG/underspeech.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    container.style.padding = '10px';
+    container.style.boxSizing = 'border-box';
+    container.style.zIndex = '100';
+    container.style.fontFamily = '"Helvetica Neue", Helvetica, Arial, sans-serif';
+    container.style.display = 'none';
+    container.style.pointerEvents = 'none';
+    return container;
+  }
+
+showDialogueFromJSON(args, util) {
```
Comment (SharkPool-SP) [inline static/extensions/PuzzlingGGG/underspeech.js:81]: util is unused

File: static/extensions/PuzzlingGGG/underspeech.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  this.dialogueContainer.innerHTML = '';
+  this.dialogueContainer.style.display = 'block';
+
+  try {
+    const dialogueData = JSON.parse(args.JSON);
+    const speakerHTML = `<div style="font-size: 1.2em; margin-bottom: 0.5em;"><strong>${args.SPEAKER}:</strong></div>`;
```
Comment (SharkPool-SP) [inline static/extensions/PuzzlingGGG/underspeech.js:87]: "args.SPEAKER" is dangerous, people can input HTML. Please use some kind of XML escape

Comment (lego7set) [reply static/extensions/PuzzlingGGG/underspeech.js:87]: oh yeah that is dangerous. do args.SPEAKER.replaceAll("&", "&amp;").replaceAll(" ", "&nbsp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;")

**Comment by PuzzlingGGG** [comment]
damn thats a lot

File: static/extensions/PuzzlingGGG/underspeech.js (hunk 1 above)
Comment (RedMan13) [reply static/extensions/PuzzlingGGG/underspeech.js:4]: vm just redirects to runtime, its better to just use whichever you have the most readily available

Comment (RedMan13) [reply static/extensions/PuzzlingGGG/underspeech.js:4]: never mind that metjod is even dummer then whay i thought

File: static/extensions/PuzzlingGGG/underspeech.js (hunk 3 above)
Comment (RedMan13) [reply static/extensions/PuzzlingGGG/underspeech.js:42]: erm actually nuhuh it adds three spaces to the category xml :trollface:

Comment (AlexSchoolOH) [reply static/extensions/PuzzlingGGG/underspeech.js:42]: > erm actually nuhuh it adds three spaces to the category xml :trollface:

only 1

Comment (RedMan13) [reply static/extensions/PuzzlingGGG/underspeech.js:42]: > > erm actually nuhuh it adds three spaces to the category xml :trollface:
> 
> only 1

notice my reference to "category XML" 💀
you can only *see* one, there are still others added to the xml though

File: static/extensions/PuzzlingGGG/underspeech.js (hunk 5 above)
Comment (RedMan13) [reply static/extensions/PuzzlingGGG/underspeech.js:87]: alternatively you can form the text into the element classes then assign innerText which will automatically escape the text for you

**Comment by RedMan13** [comment]
@PuzzlingGGG did you mean to close?

**Comment by PuzzlingGGG** [comment]
> @PuzzlingGGG did you mean to close?

its dead and I don’t want to work on it anymore

**Pull request closed without merging**